service/controller/resource/instance/delete.go

```diff
@@ -2,10 +2,133 @@ package instance
+	"sigs.k8s.io/cluster-api/util"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
 )
 
 // EnsureDeleted is a noop since the deletion of deployments is redirected to
 // the deletion of resource groups because they garbage collect them.
 func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	azureCluster, err := r.getAzureClusterFromCluster(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, *cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.removeSubnetFromAzureCluster(ctx, azureCluster, azureMachinePool.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.removeNodePool(ctx, azureMachinePool, credentialSecret)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) removeSubnetFromAzureCluster(ctx context.Context, azureCluster *capzv1alpha3.AzureCluster, subnetName string) error {
+	subnetPosition := -1
+	for i, subnet := range azureCluster.Spec.NetworkSpec.Subnets {
+		if subnet.Name == subnetName {
+			subnetPosition = i
+		}
+	}
+
+	subnetIsFound := subnetPosition >= 0
+	if subnetIsFound {
+		azureCluster.Spec.NetworkSpec.Subnets = append(azureCluster.Spec.NetworkSpec.Subnets[:subnetPosition], azureCluster.Spec.NetworkSpec.Subnets[subnetPosition+1:]...)
+
+		r.Logger.LogCtx(ctx, "message", "Ensuring subnet is not in AzureCluster", "subnetName", subnetName)
+
+		err := r.CtrlClient.Update(ctx, azureCluster)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.Logger.LogCtx(ctx, "message", "Ensured subnet is not in AzureCluster", "subnetName", subnetName)
+	}
+
+	return nil
+}
+
+func (r *Resource) removeNodePool(ctx context.Context, azureMachinePool capzexpv1alpha3.AzureMachinePool, credentialSecret *v1alpha1.CredentialSecret) error {
+	var err error
+
+	err = r.deleteARMDeployment(ctx, credentialSecret, key.ClusterID(&azureMachinePool), key.NodePoolDeploymentName(&azureMachinePool))
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.deleteVMSS(ctx, credentialSecret, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool))
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// deleteARMDeployment deletes the ARM deployment from Azure.
+func (r *Resource) deleteARMDeployment(ctx context.Context, credentialSecret *v1alpha1.CredentialSecret, resourceGroupName, deploymentName string) error {
+	r.Logger.LogCtx(ctx, "message", "Deleting machine pool ARM deployment")
+
+	deploymentsClient, err := r.ClientFactory.GetDeploymentsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if IsDeploymentNotFound(err) {
```

Comment:
Is this check in wrong place? The actual `Delete()` invocation is down below.

Reply:
lol, my bad.